### PR TITLE
[SPARK-31102][SQL] Spark-sql fails to parse when contains comment.

### PR DESCRIPTION
**Pull request by javierivanov**
### What changes were proposed in this pull request?

This PR introduces a change to false for the insideComment flag on a newline. Fixing the issue introduced by SPARK-30049.

### Why are the changes needed?

Previously on SPARK-30049 a comment containing an unclosed quote produced the following issue:
```
spark-sql> SELECT 1 -- someone's comment here
         > ;
Error in query: 
extraneous input ';' expecting <EOF>(line 2, pos 0)

== SQL ==
SELECT 1 -- someone's comment here
;
^^^
```

This was caused because there was no flag for comment sections inside the splitSemiColon method to ignore quotes. SPARK-30049 added that flag and fixed the issue, but introduced the follwoing problem:
```
spark-sql> select
         >   1,
         >   -- two
         >   2;
Error in query:
mismatched input '<EOF>' expecting {'(', 'ADD', 'AFTER', 'ALL', 'ALTER', ...}(line 3, pos 2)
== SQL ==
select
  1,
--^^^
```
This issue is generated by a missing turn-off for the insideComment flag with a newline.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

- For previous tests using line-continuity(`\`) it was added a line-continuity rule in the SqlBase.g4 file to add the functionality to the SQL context.
- A new test for inline comments was added.

### CODE REVIEW

**Comment by maropu** [comment]
ok to test

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 1]
```diff
@@ -402,24 +402,14 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
-      """SELECT concat('test', 'comment') -- someone's comment here \\
-        | comment continues here with single ' quote \\
-        | extra ' \\
-        |;""".stripMargin -> "testcomment"
-    )
-    runCliWithin(1.minute)(
-      """SELECT concat('test', 'comment') -- someone's comment here \\
-        |   comment continues here with single ' quote \\
-        |   extra ' \\
-        |   ;""".stripMargin -> "testcomment"
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Why you did you remove the existing tests instead of adding new tests?

**Comment by maropu** [comment]
cc: @wangyum

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 1 above)
Comment (javierivanov) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Hey @maropu !
The SQL parser does not recognize line-continuity per se. 
```
scala> sql(s"""SELECT concat('test', 'comment') -- someone's comment here \\\ncomment continues here with single ' quote \\\nextra ' \\""")
org.apache.spark.sql.catalyst.parser.ParseException:
mismatched input 'continues' expecting {<EOF>, ',', 'CLUSTER', 'DISTRIBUTE', 'EXCEPT', 'FROM', 'GROUP', 'HAVING', 'INTERSECT', 'LATERAL', 'LIMIT', 'ORDER', 'MINUS', 'SORT', 'UNION', 'WHERE', 'WINDOW', '-'}(line 2, pos 8)

== SQL ==
SELECT concat('test', 'comment') -- someone's comment here \
comment continues here with single ' quote \
--------^^^
extra ' \
```

It works just fine for inline comments included backslash:
```
scala> sql(s"""SELECT concat('test', 'comment') -- someone's comment here \\\n,2""") show
+---------------------+---+
|concat(test, comment)|  2|
+---------------------+---+
|          testcomment|  2|
+---------------------+---+
```

But does not work outside the inline comment(the backslash):
```
 sql(s"""SELECT concat('test', 'comment') -- someone's comment here \n,2\\\n""")
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input '\' expecting <EOF>(line 2, pos 2)

== SQL ==
SELECT concat('test', 'comment') -- someone's comment here
,2\
--^^^
``` 
Previously worked fine because of this very bug, the insideComment flag ignored everything until the end of the string. But the spark SQL parser does not recognize the backslashes. Line-continuity can be added to the CLI. But I think that feature should be added directly to the SQL parser to avoid confusion.

Let me know your thoughts 👍

Comment (maropu) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: If we can, the fix in `SqlBase.g4` (`SIMPLE_COMENT`) looks fine to me: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4#L1811 Could you try?

**Comment by maropu** [comment]
@javierivanov Any update?

**Comment by javierivanov** [comment]
@maropu I am extremly sorry, I will commit soon :)

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 1 above)
Comment (javierivanov) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:474]: @maropu I have added the fix. Let me know what you think :)

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 2]
```diff
@@ -454,24 +454,24 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
...
         |;""".stripMargin -> "testcomment"
     )
-    runCliWithin(1.minute)(
-      """SELECT concat('test', 'comment') -- someone's comment here \\
-        |   comment continues here with single ' quote \\
-        |   extra ' \\
-        |   ;""".stripMargin -> "testcomment"
-    )
   }
+
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:476]: nit: unnecessary change.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 3]
```diff
@@ -454,24 +454,24 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
...
-    runCliWithin(1.minute)(
+    runCliWithin(3.minute)(
       """SELECT concat('test', 'comment') -- someone's comment here
         |;""".stripMargin -> "testcomment"
     )
+    runCliWithin(3.minute)(
+      """SELECT concat('test', 'comment'),
+        |    -- someone's comment here
+        | 2;""".stripMargin -> "testcomment"
+    )
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:465]: Could you move this test into a new test unit just like `test("SPARK-31102: XXXX") {...}`?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 4]
```diff
@@ -454,24 +454,24 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
   }
 
   test("SPARK-30049 Should not complain for quotes in commented lines") {
-    runCliWithin(1.minute)(
+    runCliWithin(3.minute)(
```
Comment (maropu) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:457]: Why did we need this change?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 5]
```diff
@@ -1814,7 +1814,7 @@ fragment LETTER
     ;
 
 SIMPLE_COMMENT
-    : '--' ~[\r\n]* '\r'? '\n'? -> channel(HIDDEN)
+    : '--' ('\\\n' | ~[\r\n])* '\r'? '\n'? -> channel(HIDDEN)
```
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1817]: Please add tests in `PlanParserSuite`, too.

**Comment by maropu** [comment]
retest this please

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 4 above)
Comment (javierivanov) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:457]: I got timeouts running in a slow machine, changing back to 1.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
Could you check this? @wangyum

**Comment by maropu** [comment]
retest this please

**Comment by javierivanov** [comment]
ping @wangyum @maropu :)

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 5 above)
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:1817]: Ur, one more comment; could you add tests in `sql-tests/inputs/comments.sql`, too?

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
cc: @cloud-fan

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 6]
```diff
@@ -460,18 +460,20 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with BeforeAndAfterE
...
-        |;""".stripMargin -> "testcomment"
+      """SELECT concat('test', 'comment'),
+        |    -- someone's comment here
+        | 2;""".stripMargin -> "testcomment"
     )
+  }
+
+  test("SPARK-30049 Should not complain for quotes in commented with multi-lines") {
     runCliWithin(1.minute)(
-      """SELECT concat('test', 'comment') -- someone's comment here \\
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:471]: so double-slash doesn't work any more?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 7]
```diff
@@ -55,11 +55,16 @@ class PlanParserSuite extends AnalysisTest {
...
 
-  test("single comment") {
+  test("single comment case one") {
     val plan = table("a").select(star())
     assertEqual("-- single comment\nSELECT * FROM a", plan)
   }
 
+  test("single comment case two") {
+    val plan = table("a").select(star())
+    assertEqual("-- single comment\\\nwith line continuity\nSELECT * FROM a", plan)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:65]: how to interpret `\\\n`? An escaped slash and a new-line symbol?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 6 above)
Comment (javierivanov) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala:471]: It was a previous mistake since using Scala multi-line strings it auto escape chars.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala (hunk 7 above)
Comment (javierivanov) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:65]: Thats correct. Inline strings need to be escaped.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by cloud-fan** [comment]
it conflicts with 3.0, @javierivanov can you open a new PR for 3.0? Thanks!

**Comment by maropu** [comment]
@javierivanov kindly ping: https://github.com/apache/spark/pull/27920#issuecomment-627358017
